// @stepzen/sdk/lib/index.js
"use strict";
// Copyright (c) 2020,2021,2022, StepZen, Inc.
Object.defineProperty(exports, "__esModule", { value: true });
exports.init = void 0;
const path = require("path");
const authenticate_1 = require("./commands/authenticate");
const client_1 = require("./client");
const init = (sdkConfig) => {
    const defaults = {
        domain: process.env.STEPZEN_DOMAIN || 'stepzen.io',
        server: process.env.STEPZEN_SERVER_URL || 'https://{account}.stepzen.io',
    };
    return {
        verify: (account, adminkey) => {
            return authenticate_1.default({
                account,
                adminkey,
                server: defaults.server.replace('{account}', account),
                domain: defaults.domain,
            }, sdkConfig);
        },
        client: async (options) => {
            return client_1.createSdkClient(Object.assign(Object.assign({}, defaults), options), sdkConfig);
        },
    };
};
exports.init = init;
/**
 * The default SDK instance that does not know the name of the app using the SDK.
 * It tries to guess the app name from `process.argv`
 *
 * @deprecated use the init({appName: 'my-app/1.2.3'}) method to initialis an SDK instance
 */
const stepzen = exports.init({
    // For legacy apps use the script name (e.g. `index.js` as the app name)
    // Fallback to `node` if this file is imported into an interactive node shell
    appName: path.basename(process.argv[1] || process.argv[0]),
});
exports.default = stepzen;
// Hack: support both ESM-style imports: `import stepzen from '@stepzen/sdk'`
// and keep the backwards-compat with CJS-style imports: `const stepzen = require('@stepzen/sdk')`
// NOTE: if at any point in the future there is a name collision between `stepzen` and ES exports
// this hack will break
module.exports = Object.assign(Object.assign({}, module.exports), stepzen);
