// stepzen/lib/commands/whoami.js
"use strict";
// Copyright (c) 2020,2021,2022, StepZen, Inc.
Object.defineProperty(exports, "__esModule", { value: true });
const chalk = require("chalk");
const command_1 = require("@oclif/command");
const configuration_1 = require("../shared/configuration");
const utils_1 = require("../shared/utils");
const zen_command_1 = require("../shared/zen-command");
class WhoAmI extends zen_command_1.default {
    async run() {
        const { flags } = this.parse(WhoAmI);
        const configuration = await configuration_1.readConfiguration();
        if (!('account' in configuration)) {
            this.log('You are not logged in.');
            this.exit();
        }
        const details = {
            account: `${configuration.account}  `,
            adminkey: 'not set',
            apikey: 'not set',
        };
        if (flags.adminkey) {
            this.log(configuration.adminkey || details.adminkey);
            return;
        }
        if (flags.apikey) {
            this.log(configuration.apikey || details.apikey);
            return;
        }
        if (configuration.adminkey) {
            if (flags.showkeys) {
                details.adminkey = configuration.adminkey;
            }
            else {
                details.adminkey = utils_1.maskStepZenKey(configuration.adminkey);
            }
        }
        if (configuration.apikey) {
            if (flags.showkeys) {
                details.apikey = configuration.apikey;
            }
            else {
                details.apikey = utils_1.maskStepZenKey(configuration.apikey);
            }
        }
        this.log();
        this.log(`Account: ${chalk.bold(details.account)}`);
        this.log(`Admin key: ${chalk.bold(details.adminkey)}`);
        this.log(`API key: ${chalk.bold(details.apikey)}`);
        this.log();
    }
}
exports.default = WhoAmI;
WhoAmI.description = 'stepzen whoami';
WhoAmI.hidden = true;
WhoAmI.flags = Object.assign(Object.assign({}, zen_command_1.default.flags), { help: command_1.flags.help({ char: 'h' }), showkeys: command_1.flags.boolean({
        default: false,
        exclusive: ['adminkey', 'apikey'],
    }), apikey: command_1.flags.boolean({
        default: false,
        exclusive: ['showkeys', 'adminkey'],
    }), adminkey: command_1.flags.boolean({
        default: false,
        exclusive: ['showkeys', 'apikey'],
    }) });
WhoAmI.args = [];
