// stepzen/lib/shared/actions.js
"use strict";
// Copyright (c) 2020,2021,2022, StepZen, Inc.
Object.defineProperty(exports, "__esModule", { value: true });
exports.upload = exports.list = exports.deploy = void 0;
const configuration_1 = require("./configuration");
const constants_1 = require("./constants");
const stepzen_sdk_1 = require("./stepzen-sdk");
exports.deploy = async (destination, configurationsets, schema) => {
    const config = (await configuration_1.readConfiguration());
    const server = constants_1.STEPZEN_SERVER_URL.replace('{account}', config.account);
    const client = await stepzen_sdk_1.default.client({
        account: config.account,
        adminkey: config.adminkey,
        domain: constants_1.STEPZEN_DOMAIN,
        server,
    });
    const payload = { schema };
    if (configurationsets) {
        payload.configurationsets = configurationsets.split(',');
    }
    return client.deploy(destination, payload);
};
exports.list = async (type) => {
    const config = (await configuration_1.readConfiguration());
    const server = constants_1.STEPZEN_SERVER_URL.replace('{account}', config.account);
    const client = await stepzen_sdk_1.default.client({
        account: config.account,
        adminkey: config.adminkey,
        domain: constants_1.STEPZEN_DOMAIN,
        server,
    });
    return client.list[type]();
};
exports.upload = async (type, destination, source) => {
    const config = (await configuration_1.readConfiguration());
    const server = constants_1.STEPZEN_SERVER_URL.replace('{account}', config.account);
    const client = await stepzen_sdk_1.default.client({
        account: config.account,
        adminkey: config.adminkey,
        domain: constants_1.STEPZEN_DOMAIN,
        server,
    });
    return client.upload[type](destination, source);
};
