// stepzen/lib/shared/constants.js
"use strict";
// Copyright (c) 2020,2021,2022, StepZen, Inc.
Object.defineProperty(exports, "__esModule", { value: true });
exports.STEPZEN_DISCORD_URL = exports.STEPZEN_JSON2SDL_SERVER_URL = exports.ADMIN_ACCOUNT_URL = exports.ADMIN_UPLOAD_URL = exports.ADMIN_LIST_URL = exports.ADMIN_DEPLOY_URL = exports.STEPZEN_API_TEMPLATES_REPOSITORY = exports.STEPZEN_DIRECT_GENERATOR_ENGINES_URL = exports.STEPZEN_GENERATOR_ENGINES_ENDPOINT = exports.STEPZEN_GENERATOR_ENGINES_SCHEMA = exports.STEPZEN_SERVER_URL = exports.STEPZEN_DOMAIN = exports.STEPZEN_CONFIG_FILE = exports.STEPZEN_LAST_UPDATE_CHECK_TIMESTAMP = exports.STEPZEN_CONFIG_DIRECTORY = void 0;
// This file contains constants and all magic strings
const dotenv = require("dotenv");
const os = require("os");
const path = require("path");
// This allows you to set environment variables in a `.env` file.
// This file needs to be in your working directory.
dotenv.config();
const { STEPZEN_CONFIG_FILE: ENV_VAR_STEPZEN_CONFIG_FILE, STEPZEN_DOMAIN: ENV_VAR_STEPZEN_DOMAIN, STEPZEN_GENERATOR_ENGINES_SCHEMA: ENV_VAR_STEPZEN_GENERATOR_ENGINES_SCHEMA, STEPZEN_SERVER_URL: ENV_VAR_STEPZEN_SERVER_URL, STEPZEN_JSON2SDL_SERVER_URL: ENV_VAR_STEPZEN_JSON2SDL_SERVER_URL, STEPZEN_DIRECT_GENERATOR_ENGINES_URL: ENV_VAR_STEPZEN_DIRECT_GENERATOR_ENGINES_URL, } = process.env;
// Where your authentication details are stored locally
exports.STEPZEN_CONFIG_DIRECTORY = path.join(os.homedir(), '.stepzen');
exports.STEPZEN_LAST_UPDATE_CHECK_TIMESTAMP = path.join(exports.STEPZEN_CONFIG_DIRECTORY, 'last_update_check.timestamp');
exports.STEPZEN_CONFIG_FILE = ENV_VAR_STEPZEN_CONFIG_FILE || 'stepzen-config.yaml';
// The zenctl domain. Override with the env var `STEPZEN_DOMAIN`
exports.STEPZEN_DOMAIN = ENV_VAR_STEPZEN_DOMAIN || 'stepzen.io';
// The zenctl URL. Override with the env var `STEPZEN_SERVER_URL`
exports.STEPZEN_SERVER_URL = ENV_VAR_STEPZEN_SERVER_URL || 'https://{account}.stepzen.io';
// Generator Engines schema: folder/name
// Use 'stepzen/engines-dev' to test the dev instance of https://github.com/steprz/generator-engines
exports.STEPZEN_GENERATOR_ENGINES_SCHEMA = ENV_VAR_STEPZEN_GENERATOR_ENGINES_SCHEMA || 'stepzen/engines';
// Generator Engines schema: target API endpoint
exports.STEPZEN_GENERATOR_ENGINES_ENDPOINT = 'stepzen-generator/engines';
// If defined, call the generator engines cloud function directly, skipping zenctl
exports.STEPZEN_DIRECT_GENERATOR_ENGINES_URL = ENV_VAR_STEPZEN_DIRECT_GENERATOR_ENGINES_URL ||
    'https://us-central1-stepzen-functions.cloudfunctions.net/generator-engine';
// API Templates repository
exports.STEPZEN_API_TEMPLATES_REPOSITORY = 'https://github.com/steprz/stepzen-schemas';
// Paths to the endpoints used (only for tests)
exports.ADMIN_DEPLOY_URL = '/cli/admin/deploy';
exports.ADMIN_LIST_URL = '/cli/admin/list';
exports.ADMIN_UPLOAD_URL = '/cli/admin/upload';
exports.ADMIN_ACCOUNT_URL = '/cli/admin/account';
exports.STEPZEN_JSON2SDL_SERVER_URL = ENV_VAR_STEPZEN_JSON2SDL_SERVER_URL ||
    'https://jsonintrospection-ng-prod-xynkgeaaaa-uc.a.run.app';
exports.STEPZEN_DISCORD_URL = 'https://discord.gg/9k2VdPn2FR';
